homeassistant/components/mealie/services.py

```diff
@@ -70,6 +72,31 @@
     }
 )
 
+SERVICE_SET_MEALPLAN = "set_mealplan"
+SERVICE_SET_MEALPLAN_SCHEMA = vol.Any(
+    vol.Schema(
+        {
+            vol.Required(ATTR_CONFIG_ENTRY_ID): str,
+            vol.Required(ATTR_DATE): cv.date,
+            vol.Required(ATTR_ENTRY_TYPE): vol.In(
+                [x.lower() for x in MealplanEntryType]
+            ),
+            vol.Required(ATTR_RECIPE_ID): str,
+        }
+    ),
+    vol.Schema(
+        {
+            vol.Required(ATTR_CONFIG_ENTRY_ID): str,
+            vol.Required(ATTR_DATE): cv.date,
+            vol.Required(ATTR_ENTRY_TYPE): vol.In(
+                [x.lower() for x in MealplanEntryType]
+            ),
+            vol.Required(ATTR_NOTE_TITLE): str,
+            vol.Required(ATTR_NOTE_TEXT): str,
+        }
+    ),
+)
```

Comment:
Purely opinionated suggestion:
```suggestion
SERVICE_SET_MEALPLAN_SCHEMA = vol.All(
    vol.Schema(
        {
            vol.Required(ATTR_CONFIG_ENTRY_ID): str,
            vol.Required(ATTR_DATE): cv.date,
            vol.Required(ATTR_ENTRY_TYPE): vol.In(
                [x.lower() for x in MealplanEntryType]
            ),
            vol.Optional(ATTR_RECIPE_ID): object,
            vol.Optional(ATTR_NOTE_TITLE): object,
            vol.Optional(ATTR_NOTE_TEXT): object,
        }
    ),
    vol.Any(
        vol.Schema(
            {vol.Required(ATTR_RECIPE_ID): str},
            extra=vol.ALLOW_EXTRA,
        ),
        vol.Schema(
            {
                vol.Required(ATTR_NOTE_TITLE): str,
                vol.Required(ATTR_NOTE_TEXT): str,
            },
            extra=vol.ALLOW_EXTRA,
        ),
    ),
)
```
I like to split the common reqirements to a base schema and have the exclusive reqirements split out to an `Any()` block for readability.

Reply:
It should avoid introducing bugs when changing anything on the base schema - you can't only change one of them this way. Another option would be to use a constant and unpack it like
```py
SERVICE_SET_MEALPLAN_SCHEMA = vol.Any(
    vol.Schema(
        {
            **_BASE_SET_MEALPLAN_SCHEMA,
            vol.Required(ATTR_RECIPE_ID): str,
        }
    ),
```
... you get the idea.
As I said, purely opinionated 🙃